examples/vesting-wallet/tests/vesting-wallet.rs

```diff
@@ -0,0 +1,438 @@
+
+        let alice_balance = alice.wallet.get_balance(alice.address()).await?;
+        let contract_balance = alice.wallet.get_balance(contract_addr).await?;
+        let released = contract.released_0().call().await?.released;
+        let releasable = contract.releasable_0().call().await?.releasable;
+        assert_in_delta(expected_amount, released);
+        assert_in_delta(U256::ZERO, releasable);
+        assert_in_delta(
+            old_alice_balance + released
+                - U256::from(receipt.gas_used * receipt.effective_gas_price),
+            alice_balance,
+        );
+        assert_in_delta(old_contract_balance - released, contract_balance);
+
+        assert!(
+            receipt.emits(VestingWallet::EtherReleased { amount: released })
+        );
+
+        Ok(())
+    }
+
+    #[e2e::test]
+    async fn check_release_0_percent(alice: Account) -> eyre::Result<()> {
+        run_check_release(alice, 0).await
+    }
+
+    #[e2e::test]
+    async fn check_release_25_percent(alice: Account) -> eyre::Result<()> {
+        run_check_release(alice, DURATION / 4).await
+    }
+
+    #[e2e::test]
+    async fn check_release_50_percent(alice: Account) -> eyre::Result<()> {
+        run_check_release(alice, DURATION / 2).await
+    }
+
+    #[e2e::test]
+    async fn check_release_100_percent(alice: Account) -> eyre::Result<()> {
+        run_check_release(alice, DURATION).await
+    }
+
+    #[e2e::test]
+    async fn check_release_100_percent_vesting_in_the_past(
+        alice: Account,
+    ) -> eyre::Result<()> {
+        run_check_release(alice, DURATION * 4 / 3).await
+    }
+
+    #[e2e::test]
+    async fn check_vesting_schedule(alice: Account) -> eyre::Result<()> {
+        let start = block_timestamp(&alice).await?;
+        let contract_addr = alice
+            .as_deployer()
+            .with_constructor(ctr(alice.address(), start, DURATION))
+            .deploy()
+            .await?
+            .address()?;
+        let contract = VestingWallet::new(contract_addr, &alice.wallet);
+
+        let _ = watch!(contract.receiveEther().value(U256::from(BALANCE)))?;
+
+        for i in 0..64 {
+            let timestamp = i * DURATION / 60 + start;
+            let expected_amount = U256::from(std::cmp::min(
+                BALANCE,
+                BALANCE * (timestamp - start) / DURATION,
+            ));
+
+            let vested_amount =
+                contract.vestedAmount_0(timestamp).call().await?.vestedAmount;
+            assert_eq!(
+                expected_amount, vested_amount,
+                "\n---\ni: {i}\nstart: {start}\ntimestamp: {timestamp}\n---\n"
+            );
+        }
+
+        Ok(())
+    }
+}
+
+mod erc20_vesting {
+    use super::*;
+
+    async fn deploy(
+        account: &Account,
+        start: u64,
+        duration: u64,
+        allocation: u64,
+    ) -> eyre::Result<(Address, Address)> {
+        let contract_addr = account
```

Comment:
Nit: I would suggest having a dedicated fn for both eth & erc20 for deploying vesting wallet and additionally `deploy_erc20` for deploying just `erc20` token. For me it is not clear in this fn name.